examples/arrays.qasm

```diff
@@ -0,0 +1,101 @@
+OPENQASM 3.0;
+
+// ==================
+// Array declarations
+// ==================
+
+// Declare 'my_ints' to be a 1D array of 8-bit signed integers, where the index
+// runs from 0 to 15.
+array[int[8], 16] my_ints;
+
+// Declare 'my_doubles' to be a 2D array of double-precision floating-point
+// values, where the first index runs from 0 to 7 and the second from 0 to 3.
+array[float[64], 8, 4] my_doubles;
+
+// Initialise a 1D array with fixed values.
+array[uint[32], 4] my_defined_uints = {5, 6, 7, 8};
+
+// Initialise a 2D array with fixed values.
+array[float[32], 4, 2] my_defined_floats = {
+    {0.5, 0.5},
+    {1.0, 2.0},
+    {-0.4, 0.7},
+    {1.3, -2.1e-2}
+};
+
+// Initialise an array using a row of another array.  Note that this copies the
+// data; changes to the values in `my_defined_float_row` will not affect
+// `my_defined_floats`.
+array[float[32], 2] my_defined_float_row = my_defined_floats[0];
+
+// Arrays can be defined in terms of compile-time constant dimension sizes, and
+// can use expressions in the slots of their initialisers.
+const int[8] DIM_SIZE = 2;
+array[int[8], DIM_SIZE, DIM_SIZE] all_ones = {{2+3, 4-1}, {3+8, 12-4}};
```

Comment:
Does this imply an implicit cast of `DIM_SIZE` from `int[8]` to `int`?

Reply:
Do you mean when it's used as the second and third items in the array type declaration?  It's a good question - it's the same problem as asking what the type of `<SIZE>` must be in declarations like `qreg q[<SIZE>];`.  I think we haven't answered it in the types group yet.  I can change it to a generic sized `int` (actually `uint` - `int` can't be right!), but perhaps we'll allow any width in those locations; it's not something that should be an issue at runtime.

It's also on my list of examples of "stuff that needs to work for OpenQASM 2 compatibility" - you need to be able to use an integer literal with no suffix.

Reply:
Yes, I was referring to the second and third items in the array type declaration. Let's just change the `DIM_SIZE` definition to `uint` and we can address widths later.